File: Homepwner/Homepwner/BNRItemStore.m
//
//  BNRItemStore.m
//  Homepwner
//
//  Created by Kevin Lin on 4/18/14.
//  Copyright (c) 2014 Big Nerd Ranch. All rights reserved.
//

#import "BNRItemStore.h"
#import "BNRImageStore.h"

@interface BNRItemStore ()

@property (nonatomic) NSMutableArray *privateItems;

@end


@implementation BNRItemStore

+ (instancetype)sharedStore
{
    static BNRItemStore *sharedStore = nil;
    if (!sharedStore) {
        sharedStore = [[BNRItemStore alloc] initPrivate];
    }
    
    return sharedStore;
}

- (instancetype)init
{
    @throw [NSException exceptionWithName:@"Singleton"
                                   reason:@"Use +[BNRItemStore sharedStore]"
                                 userInfo:nil];
    return nil;
}

- (instancetype)initPrivate
{
    self = [super init];
    
    if (self) {
        NSString *path = [self itemArchivePath];
        _privateItems = [NSKeyedUnarchiver unarchiveObjectWithFile:path];
        
        if (!_privateItems) {
            _privateItems = [[NSMutableArray alloc] init];
        }
    }
    
    return self;
}

- (NSArray *)allItems
{
    return self.privateItems;
}

- (BNRItem *)createItem
{
    BNRItem *item = [[BNRItem alloc] init];
    [self.privateItems addObject:item];
    return item;
}

- (void)removeItem:(BNRItem *)item
{
    [[BNRImageStore sharedStore] deleteImageForKey:item.itemKey];
    [self.privateItems removeObjectIdenticalTo:item];
}

- (void)moveItemAtIndex:(NSUInteger)fromIndex
                toIndex:(NSUInteger)toIndex
{
    if (fromIndex == toIndex) {
        return;
    }
    
    BNRItem *item = self.privateItems[fromIndex];
    [self.privateItems removeObjectIdenticalTo:item];
    [self.privateItems insertObject:item atIndex:toIndex];
}

- (BOOL)saveChanges
{
    NSString *path = [self itemArchivePath];
    
    return [NSKeyedArchiver archiveRootObject:self.privateItems toFile:path];
}

- (NSString *)itemArchivePath
{
    NSArray *documentDirectories = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
    NSString *documentDirectory = [documentDirectories firstObject];
    
    return [documentDirectory stringByAppendingPathComponent:@"items.archive"];
}

@end

File: Homepwner/Homepwner/BNRItemStore.swift
//
//  BNRItemStore.swift
//  Homepwner
//
//  Created by Kevin Lin on 6/3/14.
//  Copyright (c) 2014 Big Nerd Ranch. All rights reserved.
//

import Foundation

var _sharedBNRItemStore : BNRItemStore?

@objc class BNRItemStore: NSObject {
    var privateItems : NSMutableArray
    
    init(isPrivate: Bool) {
        let path = BNRItemStore.itemArchivePath()
        let storedItems = NSKeyedUnarchiver.unarchiveObjectWithFile(path) as? NSMutableArray
        
        if let items = storedItems {
            self.privateItems = items
        } else {
            self.privateItems = NSMutableArray()
        }
        
        super.init()
    }
    
    class func sharedStore() -> BNRItemStore {
        if _sharedBNRItemStore == nil {
            _sharedBNRItemStore = BNRItemStore(isPrivate: true)
        }
        
        return _sharedBNRItemStore!
    }
    
    func allItems() -> NSArray {
        return privateItems
    }
    
    func createItem() -> BNRItem {
        let item = BNRItem(random: true)
        privateItems.addObject(item)
        return item
    }
    
    func removeItem(item: BNRItem) {
        //let imageStore = BNRImageStore.sharedStore()
        //imageStore.deleteImageForKey(item.itemKey)
        privateItems.removeObject(item)
    }
    
    func moveItemAtIndex(fromIndex: Int, toIndex: Int) {
        if (fromIndex == toIndex) {
            return
        }
        
        let item = privateItems[fromIndex] as! BNRItem
        privateItems.removeObjectIdenticalTo(item)
        privateItems.insertObject(item, atIndex: toIndex)
    }
    
    func saveChanges() -> Bool {
        let path = BNRItemStore.itemArchivePath();
        
        return NSKeyedArchiver.archiveRootObject(self.privateItems, toFile: path)
    }
    
    class func itemArchivePath() -> String {
        let documentDirectories = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true)
        let documentDirectory = documentDirectories[0] as! String
        
        return documentDirectory.stringByAppendingPathComponent("items.archive")
    }

}

